src/process.rs

```diff
@@ -0,0 +1,130 @@
+/// Select the best match in a list or dictionary of choices.
+///
+/// Find best matches in a list or dictionary of choices, return a generator of tuples containing
+/// the match and its score. If a dictionary is used, also returns the key for each match.
+///
+/// TODO: Add support for choices as HashMap<&str, &str>, not only as slice &[&str].
+pub fn extract_without_order<'a, I, T>(
+    query: &str,
+    choices: I,
+    processor: &dyn Fn(&str, bool) -> String,
```

Comment:
my rust is rusty so I'm not prepared to make a declaration on this without experiments. however, have you considered not using trait objects? I tested it and it works without modification to the call sites. I need to run some tests to see if its worth it.

We can also just leave this as is, right now. The crate is apparently published by someone that isn't me and I need to figure out how to get in touch with them, so even if changing this to a non-trait object later is a breaking change, we can probably do that before the next crate push.

Reply:
I believe the main benefit of using trait objects is that arguments to `processor` can be closures which capture their environment (i.e. receive some sort of reference to `self`).

e.g.

Capturing environment with a closure:
```rust
fn main() {
    let x = 4;
    let equal_to_x = |z| z == x;
    let y = 4;
    assert!(equal_to_x(y));
}
```

With a function:
```rust
fn main() {
    let x = 4;
    fn equal_to_x(z: i32) -> bool {
        z == x
    }
    let y = 4;
    assert!(equal_to_x(y));
}

$ cargo run
   Compiling equal-to-x v0.1.0 (file:///projects/equal-to-x)
error[E0434]: can't capture dynamic environment in a fn item
 --> src/main.rs:5:14
  |
5 |         z == x
  |              ^
  |
  = help: use the `|| { ... }` closure form instead

error: aborting due to previous error

For more information about this error, try `rustc --explain E0434`.
error: could not compile `equal-to-x`.

To learn more, run the command again with --verbose.
```

Using the trait object [`Fn`](https://doc.rust-lang.org/std/ops/trait.Fn.html) allows for arguments to be closures which capture their environment, while [primitive function pointers](https://doc.rust-lang.org/std/primitive.fn.html) do not support this type of argument.

Reply:
I think this is untrue. With your branch checked out, changing the signature for processor to `U ... where U: U: Fn(&str, bool) -> String` in `extract_one` and `extract_without_order` allows me to do the following in `unwrap_extract_one_choice`:

```rust
let b = true;
extract_one(
    query,
    get_baseball_strings().iter(),
    |a, _| utils::full_process(a, b),
    &fuzz::wratio,
    0)
... etc
```

It allows: a) function references b) functions by value (i.e., `utils::full_process` without `&`) and c) closures with or without captures.

If you don't want to change it, feel free to resolve this. But I will look into making the change before publishing the crate.

Reply:
Okay, I've replaced the usages of trait objects with trait bounds. There is no compelling reason to use trait objects here, it would seem.